vortex-array/src/array/list/mod.rs

```diff
@@ -0,0 +1,288 @@
+mod compute;
+
+use std::fmt::Display;
+use std::sync::Arc;
+
+use num_traits::AsPrimitive;
+use serde::{Deserialize, Serialize};
+use vortex_dtype::{match_each_native_ptype, DType};
+use vortex_error::{vortex_bail, vortex_panic, VortexExpect, VortexResult};
+
+use crate::array::{NullArray, PrimitiveArray};
+use crate::compute::{scalar_at, slice};
+use crate::encoding::ids;
+use crate::stats::{ArrayStatistics, Stat, StatisticsVTable, StatsSet};
+use crate::validity::{LogicalValidity, Validity, ValidityMetadata, ValidityVTable};
+use crate::variants::{ListArrayTrait, PrimitiveArrayTrait, VariantsVTable};
+use crate::visitor::{ArrayVisitor, VisitorVTable};
+use crate::{
+    impl_encoding, ArrayDType, ArrayData, ArrayLen, ArrayTrait, Canonical, IntoArrayData,
+    IntoArrayVariant, IntoCanonical,
+};
+
+impl_encoding!("vortex.list", ids::LIST, List);
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct ListMetadata {
+    validity: ValidityMetadata,
+    element_len: usize,
+    offset_dtype: DType,
+}
+
+impl Display for ListMetadata {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "ListMetadata")
+    }
+}
+
+// A list is valid if the:
+// - offsets start at a value in elements
+// - offsets are sorted
+// - the final offset points to an element in the elements list, pointing to zero
+//   if elements are empty.
+// - final_offset >= start_offset
+// - The size of the validity is the size-1 of the offset array
+
+impl ListArray {
+    pub fn try_new(
+        elements: ArrayData,
+        offsets: ArrayData,
+        validity: Validity,
+    ) -> VortexResult<Self> {
+        let nullability = validity.nullability();
+        let list_len = offsets.len() - 1;
+        let element_len = elements.len();
+        let offset_dtype = offsets.dtype().clone();
+
+        let validity_metadata = validity.to_metadata(list_len)?;
+
+        if !offsets.dtype().is_int() || offsets.dtype().is_nullable() {
+            vortex_bail!(
+                "Expected offsets to be an non-nullable integer type, got {:?}",
+                offsets.dtype()
+            );
+        }
+
+        if offsets.is_empty() {
+            vortex_bail!("Offsets must have at least one element, [0] for an empty list");
+        }
+
+        let Some(min_value) = offsets.statistics().compute_as_cast::<i64>(Stat::Min) else {
```

Comment:
You shouldn't compute a statistic here, constructing an array should be relatively cheap. You can force a signed integer offsets dtype if you like

Reply:
I added an unchecked path